core/corehttp/gateway_handler.go

```diff
@@ -148,13 +154,11 @@ func (i *gatewayHandler) getOrHeadHandler(w http.ResponseWriter, r *http.Request
 
 	// set these headers _after_ the error, for we may just not have it
 	// and dont want the client to cache a 500 response...
-	// and only if it's /ipfs!
-	// TODO: break this out when we split /ipfs /ipns routes.
+	w.Header().Set("ETag", etag)
+	w.Header().Set("Cache-Control", fmt.Sprintf("public, max-age=%d", int(res.TTL.Seconds())))
```

Comment:
nope, errors still happen below. this is trickty! --  maybe the error functions should _clear_ this header to be sure.